hook using node doesn't work on FreeBSD

We have a hook using doctoc, the node package. I cannot install the hook, this is the output:

```
desktop1% pre-commit install --install-hooks                                                                                                                                                                                                  
pre-commit installed at .git/hooks/pre-commit
[INFO] Initializing environment for git@github.com:jockej/pre-commit-hooks.git.
[INFO] Initializing environment for https://github.com/thlorenz/doctoc.
[INFO] Initializing environment for https://gitlab.com/pycqa/flake8.
[INFO] Initializing environment for https://github.com/jorisroovers/gitlint.
[INFO] Installing environment for https://github.com/thlorenz/doctoc.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/bin/node', '/usr/local/bin/npm', 'install', '-g', '.')
return code: 254
expected return code: 0
stdout: (none)
stderr:
    npm ERR! code ENOENT
    npm ERR! syscall chmod
    npm ERR! path /home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/lib/node_modules/doctoc/doctoc.js
    npm ERR! errno -2
    npm ERR! enoent ENOENT: no such file or directory, chmod '/home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/lib/node_modules/doctoc/doctoc.js'
    npm ERR! enoent This is related to npm not being able to find a file.
    npm ERR! enoent 
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/joakim/.npm/_logs/2020-05-07T10_30_16_218Z-debug.log
    
Check the log at /home/joakim/.cache/pre-commit/pre-commit.log
```
These hooks all install and run perfectly on GNU/Linux.

I believe it happens because pre-commit tries to use the --prebuilt flag to nodeenv, however this doesn't work for FreeBSD since there are no prebuilt packages (the URL constructed gives a 404). I have node installed so it should be possible to use --system, but I haven't been able to make it work myself.

Nodeenv itself seems to work fine with --system.

Please let me know if I can help with trying things on FreeBSD.

Strange! I get the same error in a freshly created repo. (I forgot to cat the files in the beginning, they are at the end)
```
desktop1% cd test 
desktop1% ls
desktop1% mkdir test-pre-commit
desktop1% cd test-pre-commit 
desktop1% git init
Initialized empty Git repository in /usr/home/joakim/test/test-pre-commit/.git/
desktop1% pre-commit install
pre-commit installed at .git/hooks/pre-commit
desktop1% pre-commit run --all-files
[INFO] Installing environment for https://github.com/thlorenz/doctoc.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/bin/node', '/usr/local/bin/npm', 'install', '-g', '.')
return code: 254
expected return code: 0
stdout: (none)
stderr:
    npm ERR! code ENOENT
    npm ERR! syscall chmod
    npm ERR! path /home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/lib/node_modules/doctoc/doctoc.js
    npm ERR! errno -2
    npm ERR! enoent ENOENT: no such file or directory, chmod '/home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/lib/node_modules/doctoc/doctoc.js'
    npm ERR! enoent This is related to npm not being able to find a file.
    npm ERR! enoent 
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/joakim/.npm/_logs/2020-05-08T12_00_15_940Z-debug.log
    
Check the log at /home/joakim/.cache/pre-commit/pre-commit.log
desktop1% cat .pre-commit-config.yaml 
repos:
  - repo: https://github.com/thlorenz/doctoc
    rev: v1.4.0
    hooks:
      - id: doctoc
        # Assume second level heading and GitHub flavouring.
        args: [--title=## Contents, --github]
        stages: [commit]
desktop1% cat README.md 
hejsan

hej hej
desktop1% 
```

Plain nodeenv seems to work fine.
```
desktop1% echo $HOME            
/home/joakim
desktop1% nodeenv nenv -n system
desktop1% . nenv/bin/activate
(nenv) desktop1% npm install -g doctoc
/usr/home/joakim/nenv/bin/doctoc -> /usr/home/joakim/nenv/lib/node_modules/doctoc/doctoc.js
+ doctoc@1.4.0
added 67 packages from 41 contributors in 6.285s
(nenv) desktop1% 
```
It has to be something about how pre-commit calls nodeenv, no? When it works it uses the path /usr/home/joakim (like above), when it doesn't, it seems to be using /home/joakim:

```
An unexpected error has occurred: CalledProcessError: command: ('/home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/bin/node', '/usr/local/bin/npm', 'install', '-g', '.')
return code: 254
expected return code: 0
stdout: (none)
stderr:
    npm ERR! code ENOENT
    npm ERR! syscall chmod
    npm ERR! path /home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/lib/node_modules/doctoc/doctoc.js
    npm ERR! errno -2
    npm ERR! enoent ENOENT: no such file or directory, chmod '/home/joakim/.cache/pre-commit/repoizbdwm4g/node_env-system/lib/node_modules/doctoc/doctoc.js'
    npm ERR! enoent This is related to npm not being able to find a file.
    npm ERR! enoent 
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/joakim/.npm/_logs/2020-05-08T12_00_15_940Z-debug.log
    
Check the log at /home/joakim/.cache/pre-commit/pre-commit.log
```

and here's a reproduction using only node and npm on linux:

```dockerfile
FROM ubuntu:focal

RUN : \
    && apt-get update \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
        ca-certificates \
        curl \
        git \
        gnupg2 \
        software-properties-common \
    && curl --silent https://deb.nodesource.com/gpgkey/nodesource.gpg.key | apt-key add - \
    && add-apt-repository https://deb.nodesource.com/node_12.x \
    && apt-get update \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
        nodejs \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

# simulate the home layout of freebsd
RUN : \
    && rmdir /home \
    && mkdir -p /usr/home \
    && ln -s /usr/home /home

RUN : \
    && useradd -d /home/asottile asottile \
    && mkdir -p /home/asottile \
    && chown asottile /home/asottile
USER asottile

# cd before each command is chosen instead of `WORKDIR` as the latter
# dereferences symlinks

RUN : \
    && cd /home/asottile \
    && git clone https://github.com/thlorenz/doctoc doctoc \
    && cd doctoc \
    && git checkout v1.4.0

# this simulates how `nodeenv` builds a separate environment
ENV \
    NODE_VIRTUAL_ENV=/home/asottile/doctoc/nenv \
    NPM_CONFIG_PREFIX=/home/asottile/doctoc/nenv \
    npm_config_prefix=/home/asottile/doctoc/nenv \
    NODE_PATH=/home/asottile/doctoc/nenv/lib/node_modules \
    PATH=/home/asottile/doctoc/nenv/bin:$PATH

RUN : \
    && cd /home/asottile/doctoc \
    && /usr/bin/node /usr/bin/npm install \
    && /usr/bin/node /usr/bin/npm install -g . || ( \
        # adjust this to look at the failures
        ls -al "$NODE_PATH" && \
        exit 1\
    )
```

```console
$ docker build -t test .
...
Step 8/8 : RUN :     && cd /home/asottile/doctoc     && /usr/bin/node /usr/bin/npm install     && /usr/bin/node /usr/bin/npm install -g . || (         ls -al "$NODE_PATH" &&         exit 1    )
 ---> Running in f9d8526451ae
npm WARN deprecated request@2.79.0: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated hawk@3.1.3: This module moved to @hapi/hawk. Please make sure to switch over as this distribution is no longer supported and may contain bugs and critical security issues.
npm WARN deprecated hoek@2.16.3: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated cryptiles@2.0.5: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated boom@2.10.1: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).
npm WARN deprecated sntp@1.0.9: This module moved to @hapi/sntp. Please make sure to switch over as this distribution is no longer supported and may contain bugs and critical security issues.
npm WARN deprecated core-js@2.6.11: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.
npm notice created a lockfile as package-lock.json. You should commit this file.
added 365 packages from 251 contributors and audited 793 packages in 21.694s

21 packages are looking for funding
  run `npm fund` for details

found 58 vulnerabilities (26 low, 7 moderate, 24 high, 1 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
npm ERR! code ENOENT
npm ERR! syscall chmod
npm ERR! path /home/asottile/doctoc/nenv/lib/node_modules/doctoc/doctoc.js
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, chmod '/home/asottile/doctoc/nenv/lib/node_modules/doctoc/doctoc.js'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/asottile/.npm/_logs/2020-05-08T20_13_59_212Z-debug.log
total 8
drwxr-xr-x 2 asottile asottile 4096 May  8 20:13 .
drwxr-xr-x 3 asottile asottile 4096 May  8 20:13 ..
lrwxrwxrwx 1 asottile asottile   42 May  8 20:13 doctoc -> ../../../../../../usr/home/asottile/doctoc
The command '/bin/sh -c :     && cd /home/asottile/doctoc     && /usr/bin/node /usr/bin/npm install     && /usr/bin/node /usr/bin/npm install -g . || (         ls -al "$NODE_PATH" &&         exit 1    )' returned a non-zero code: 1
```

--- a/pre_commit/store.py
+++ b/pre_commit/store.py
@@ -30,10 +30,11 @@ def _get_default_directory() -> str:
     `Store.get_default_directory` can be mocked in tests and
     `_get_default_directory` can be tested.
     """
-    return os.environ.get('PRE_COMMIT_HOME') or os.path.join(
+    ret = os.environ.get('PRE_COMMIT_HOME') or os.path.join(
         os.environ.get('XDG_CACHE_HOME') or os.path.expanduser('~/.cache'),
         'pre-commit',
     )
+    return os.path.realpath(ret)
 
 
 class Store:

--- a/tests/store_test.py
+++ b/tests/store_test.py
@@ -25,7 +25,8 @@ def test_our_session_fixture_works():
 def test_get_default_directory_defaults_to_home():
     # Not we use the module level one which is not mocked
     ret = _get_default_directory()
-    assert ret == os.path.join(os.path.expanduser('~/.cache'), 'pre-commit')
+    expected = os.path.realpath(os.path.expanduser('~/.cache/pre-commit'))
+    assert ret == expected
 
 
 def test_adheres_to_xdg_specification():
@@ -33,7 +34,8 @@ def test_adheres_to_xdg_specification():
         os.environ, {'XDG_CACHE_HOME': '/tmp/fakehome'},
     ):
         ret = _get_default_directory()
-        assert ret == os.path.join('/tmp/fakehome', 'pre-commit')
+    expected = os.path.realpath('/tmp/fakehome/pre-commit')
+    assert ret == expected
 
 
 def test_uses_environment_variable_when_present():
@@ -41,7 +43,8 @@ def test_uses_environment_variable_when_present():
         os.environ, {'PRE_COMMIT_HOME': '/tmp/pre_commit_home'},
     ):
         ret = _get_default_directory()
-        assert ret == '/tmp/pre_commit_home'
+    expected = os.path.realpath('/tmp/pre_commit_home')
+    assert ret == expected
 
 
 def test_store_init(store):

